module/ice-9/generic-hash-tables.scm

```diff
@@ -0,0 +1,854 @@
+            ;; Mapping and folding
+            hash-table-map hash-table-for-each hash-table-map! hash-table-map->list
+            hash-table-fold hash-table-prune!
+            ;; Copying
+            hash-table-copy hash-table-empty-copy
+            ;; Conversion
+            hash-table->alist
+            ;; Hash tables as sets
+            hash-table-union! hash-table-intersection! hash-table-difference!
+            hash-table-xor!
+            ;; Hashing
+            string-ci-hash hash-by-identity hash-by-value hash)
+  #:re-export (string-hash)
+  #:replace (hash make-hash-table hash-table?))
+
+
+;;;; Internal helper macros
+
+;; Define these first, so the compiler will pick them up.
+
+;; I am a macro only for efficiency, to avoid varargs/apply.
+(define-macro (hashx-invoke hashx-proc ht-var . args)
+  "Invoke HASHX-PROC, a `hashx-*' procedure taking a hash-function,
+assoc-function, and the hash-table as first args."
+  `(,hashx-proc (ht-hash-function ,ht-var)
+		(ht-associator ,ht-var)
+		(ht-real-table ,ht-var)
+		. ,args))
+
+(define-macro (with-hashx-values bindings ht-var . body-forms)
+  "Bind BINDINGS to the hash-function, associator, and real-table of
+HT-VAR, while evaluating BODY-FORMS."
+  `(let ((,(third bindings) (ht-real-table ,ht-var))
+         (,(first bindings) (ht-hash-function ,ht-var))
+	 (,(second bindings) (ht-associator ,ht-var)))
+     . ,body-forms))
+
+(define-syntax assert-mutable
+  (syntax-rules ()
+    ((assert-mutable ht)
+     (or (hash-table-mutable? ht)
+         (error "Hash table is not mutable" ht)))))
+
+
+;;;; Hashing
+
+;;; SRFI-125 and R6RS hash functions are supposed to accept only one
+;;; argument, but Guile standard hash tables needs two args.
+;;; Therefore, the hash functions inside the hash table always accepts
+;;; one (required) argument and at least one possible argument, which
+;;; must be a fixnum.
+
+;;; The largest fixnum is in `most-positive-fixnum' in module (guile),
+;;; though not documented anywhere but libguile/numbers.c.
+
+(define (caller-with-default-size hash-fn)
+  "Answer a function that makes `most-positive-fixnum' the default
+second argument to HASH-FN, a 2-arg procedure."
+  (lambda* (obj #:optional (size most-positive-fixnum))
+    (hash-fn obj size)))
+
+(define hash (caller-with-default-size (@ (guile) hash)))
+
+(define string-ci-hash string-hash-ci)
+
+(define hash-by-identity (caller-with-default-size hashq))
+(define hash-by-value    (caller-with-default-size hashv))
+
+(define (wrap-hash-function hash-function)
+  (lambda* (obj size)
+    (modulo (hash-function obj) size)))
+
+
+;;;; Reflective queries, construction, predicate
+
+(define (get-hash-functions equiv-function hash-function)
+  "Returns an internal and an external hash function."
+  (cond
+   (hash-function (cond
+                   ;; SRFI-69 should give HASH as default hash-function.
+                   ((or (eq? (@ (guile) hash) hash-function)
```

Comment:
`(@ (guile) hash)` is used several time through the code, maybe it is worth defining it as `(define guile-hash (@ (guile) hash))` somehere to the top of the file.

Reply:
Imported `guile:hash`, `guile:assoc` and `guile:make-hash-table` now.